scripts/dns-any-query.nse

```diff
@@ -0,0 +1,159 @@
+local stdnse = require "stdnse"
+local dns = require "dns"
+local shortport = require "shortport"
+local target = require "target"
+
+description = [[
+Checks if ANY dns type is allowed.
+
+Usually if server allows ANY queries, it will send more than one answers.
+During prerule phase, if <code>newtargets</code> script argument is passed
+it will find new targets to nmap scanning queue
+]]
+
+---
+-- @usage
+-- nmap --script=dns-any-query -p53 <host>
+--
+-- Script uses dns library to connect to dns server
+--
+-- @output
+-- PORT   STATE SERVICE REASON
+-- 53/udp open  domain  udp-response ttl 55
+-- | dns-any-query:
+-- |   Result: Server is likely responding to ANY query
+-- |   Authoritative: TRUE
+-- |   Found:
+-- |     		MX
+-- |     		SOA
+-- |     		TXT
+-- |     		TXT
+-- |_    		TXT
+
+author = "Erhad Husovic"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"default","safe"}
+
+local options = {}
+
+-- During prerule phase, we except two variables: dns-any-query.server to query against and
+-- dns-any-query.query-domain domain to query
+-- Also, passing additional script arg <code>newtargets</code> will add domain name and server
+-- to new targets
+prerule = function()
+  options.domain_name, options.server, options.full_output,options.newtargets = stdnse.get_script_args(
+	"dns-any-query.query-domain","dns-any-query.server","dns-any-query.full-output","newtargets"
+  )
+
+  if not options.domain_name then
+    return false
+  end
+
+  if not options.server then
+    return false
+  end
+
+  options.port = 53
+  if options.newtargets then
+    local status, err = target.add(options.domain_name,options.server)
+  end
+  return true
+end
+
+-- During hostrule phase, we check if there is dns-any-query.server passed as argument
+-- If argument isn't passed, we will set server as host.targetname or host.name
+-- false is returned if neither of them is accessible
+hostrule = function(host)
+  options.server = stdnse.get_script_args("dns-any-query.server")
+  options.port = 53
+
+  if not host.name then
+    if not host.targetname then
+      return false
+    else
+     options.domain_name = host.target
+    end
+  else
+    options.domain_name = host.name
+  end
+
+  if not options.server then
+    return false
+  end
+
+  return true
+end
+
+-- portrule checks if we received 53/udp port and tries to obtain query-domain from
+-- script arguments.
+-- if query-domain isn't passed, we try to extract domain name from parts of host 
+portrule = function(host,port)
+  stdnse.debug1('trigerred portrule')
+  if shortport.portnumber(53,'udp')(host,port) then
```

Comment:
DNS can and does listen on TCP, see the [dns-nsid](https://github.com/nmap/nmap/blob/master/scripts/dns-nsid.nse#L45-L57) script for how to handle tcp/udp.

Reply:
yeah, forgot to add that part